### PR TITLE
Off days setup

### LINKED ISSUE #8: Off days
Create table `off_days` with columns:
`id, user_id, company_id, from, to, type, approved`

This will be a table for vacation days and sick days, which are declared by employee, and later approved by company admin.

### PR DESCRIPTION
**Pull request by KenzR**
Closes #8 
Created off days page

### CODE REVIEW

File: app/views/off_days/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+      <ul>
+      <% @off_day.errors.full_messages.each do |message| %>
+        <li><%= message %></li>
+      <% end %>
+      </ul>
+    </div>
+  <% end %>
+
+  <div class="field">
```
Comment (sls-lnluksa) [inline app/views/off_days/_form.html.erb]: Remove all hidden fields!
We don't need to send any company_id or user_id. That is all accessible in controller.

Add dropdown for company admin. Drop down will have list of all employees so admin can chose to request off days for someone.
Handle this in controller.

File: db/migrate/20160323141142_create_off_days.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class CreateOffDays < ActiveRecord::Migration
+  def change
+    create_table :off_days do |t|
+      t.integer :user_id
+      t.integer :company_id
+      t.date :from
+      t.date :to
+      t.string :absence_type
+      t.boolean :approved, default: false
+
```
Comment (sls-lnluksa) [inline db/migrate/20160323141142_create_off_days.rb]: Add fields requested_by and approved_by

File: app/views/off_days/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    <%= f.hidden_field :user_id %><br>
+    <%= f.hidden_field :user_id, value: current_user.id %>
+  </div>
+  <div class="field">
+    <%= f.hidden_field :company_id %><br>
+    <%= f.hidden_field :company_id, value: current_company.id %>
+  </div>
+  <div class="field">
+    <%= f.label :from %><br>
+    <%= f.text_field :from, "data-provide" => 'datepicker'%>
```
Comment (sls-lnluksa) [inline app/views/off_days/_form.html.erb]: Format date to 'YYYY-MM-DD'
That should fix saving problems

File: app/models/off_day.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+class OffDay < ActiveRecord::Base
```
Comment (sls-lnluksa) [inline app/models/off_day.rb]: Add list of possible off_days. This list should populate "type" in select. User should only chose from those, not to type whatever he likes.

File: app/views/off_days/_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  </div>
+  <div class="field">
+    <%= f.label :to %><br>
+    <%= f.text_field :to, "data-provide" => 'datepicker'%>
+  </div>
+  <div class="field">
```
Comment (sls-lnluksa) [inline app/views/off_days/_form.html.erb]: Select from predefined values.

File: app/policies/off_day_policy.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  def update?
+    @current_user.current_company_rights.company_admin? || @current_user.current_company_rights.employee?
+  end
+
+  def destroy?
+    return false if @current_user == @user
+    @current_user.current_company_rights.company_admin?
+  end
+
+  def edit?
```
Comment (sls-lnluksa) [inline app/policies/off_day_policy.rb]: edit and update
should be possible to access by employee only if record is not approved.
After that, record can be edited by company_admin.

File: app/views/off_days/admin_panel.html.erb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        <%end%>
+        </td>
+      </tr>
+    <% end %>
+  </tbody>
+</table>
+
+<br>
+
+<%= link_to 'New Off day', new_off_day_path %>
```
Comment (IvanPavlovic92) [inline app/views/off_days/admin_panel.html.erb]: There should be one view for this no need for separate view for admin.
if statement is enough for this

File: app/views/off_days/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,41 @@
+<p id="notice"><%= notice %></p>
+
+<h1>Listing Off Days</h1>
+
+<table border="1">
```
Comment (IvanPavlovic92) [inline app/views/off_days/index.html.erb]: Table should be only displayed if there is some data

Comment (IvanPavlovic92) [inline app/views/off_days/index.html.erb]: If something isn't input correctly corresponding notes should be flashed.
Validations for off-day form should be done in front end as well.

File: app/views/layouts/_header.html.erb [hunk 9]
```diff
@@ -30,6 +30,7 @@
               <% if current_user.admin? %>
                 <li><%= link_to "Users", users_path %> </li>
                 <li><%= link_to "Companies", companies_path %> </li>
+                <li><%= link_to "Off days panel", off_days_path %></li>
```
Comment (IvanPavlovic92) [inline app/views/layouts/_header.html.erb]: Shouldn't be a regular user be able to request an off day for himself,
especially if we have admin and regular view for off days page?

File: app/views/off_days/index.html.erb (hunk 8 above)
Comment (IvanPavlovic92) [inline app/views/off_days/index.html.erb]: if you create new off day as an employee you get redirected to show page,
for which the access is denied and eventually redirected to same new page,
but the off day has been created.

It should instead redirect you to show page.

User should be able to see his individual off day and edit it or destroy it unless it is approved.

File: app/controllers/off_days_controller.rb [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    @reasons = OffDay.reasons
+    @company_employees = current_user.current_company.users
+    authorize @off_day
+  end
+
+  def edit
+    authorize @off_day
+    @company_employees = current_user.current_company.users
+    if @off_day.approved == true && current_user.current_company_rights.employee?
+      redirect_to off_days_path, alert: 'Off days was approved, u have no more access to edit page'
```
Comment (IvanPavlovic92) [inline app/controllers/off_days_controller.rb]: Could not reproduce this

File: app/controllers/off_days_controller.rb [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    authorize @off_day
+    @off_day.destroy
+    respond_to do |format|
+      format.html { redirect_to off_days_admin_url, notice: 'Off day was successfully destroyed.' }
+      format.json { head :no_content }
+    end
+  end
+
+  def approve
+    off_day = OffDay.find(params[:id])
```
Comment (IvanPavlovic92) [inline app/controllers/off_days_controller.rb]: Add this method to before filter set_off_day and use @off_day

File: app/controllers/off_days_controller.rb [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    authorize off_day
+    off_day.approved = true
+    off_day.approved_by = current_user.id
+    if off_day.save
+      redirect_to off_days_admin_url, notice: 'Off day was successfully approved.'
+    else
+      redirect_to off_days_admin_url, alert: 'Off day was unsuccesfully approved'
```
Comment (IvanPavlovic92) [inline app/controllers/off_days_controller.rb]: This alert is a confusing sentence.

File: app/controllers/off_days_controller.rb [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    end
+  end
+
+  def admin_panel
```
Comment (IvanPavlovic92) [inline app/controllers/off_days_controller.rb]: Not needed if we have only one view for index.

File: app/models/off_day.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+  valid_date_regex = /\d{4}\-\d{2}\-\d{2}/
+  validates :from, presence: true
+  validates :to, presence: true
+  validates_format_of :from, :with => valid_date_regex, :alert => 'Date must be in valid format'
```
Comment (IvanPavlovic92) [inline app/models/off_day.rb]: Alert is not showing anywhere

File: app/views/off_days/_form.html.erb [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  </div>
+  <div class="actions">
+    <%= f.submit %>
+  </div>
+<% end %>
+
+<%= javascript_tag do %>
+  $(document).ready(function(){
+    $(function() {
+      $( "#off_day_from" ).datepicker({ dateFormat: 'yy-mm-dd' });
```
Comment (IvanPavlovic92) [inline app/views/off_days/_form.html.erb]: instead of id use class, so we can only add one class to all inputs where we need datepicker

**Pull request merged**